VPLAY-10090 VPLAY-10399 - Audio loss in Trick to play transition in L2 tests

anjali-syna opened the pull request:
Reason for change: Audio is lost while performing live pause  and fast forward to live using LLD test asset in L2 test environment.
Changes: Perform Flush after Configure in TuneHelper. Also, call flush with shouldTearDown false, because in case of a new tune, pipeline might not be in a playing/paused state which causes Flush() to destroy the pipeline. This has to be avoided.
Test Procedure: Refer Jira Ticket

anjali-syna committed:
VPLAY-10090 [ENT-4082] Audio loss in Trick to play transition in L2 tests

Reason for change: Audio is lost while performing fast forward to live
 in L2 test environment. This is because in TuneHelper,  Configure is
 happening after Flush() call with proper seek position.
 This results in pipeline loosing the flush and seek, thus affecting AV sync.

Changes:
* Perform Flush after Configure in TuneHelper
* Also, call flush with shouldTearDown false, because in case of a new tune,
  pipeline might not be in a playing/paused state which causes Flush() to
  destroy the pipeline. This has to be avoided.
* Adding new unit test:TuneHelperWithAampTsbConfigureFlushSequence

Test Procedure: Run L2 tests

Risks: Low

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0449a0ba518cd72585823907892ca556
PRESENT
-->

# Coverity Issue - Data race condition
Accessing "pMediaStreamContext->numberOfFragmentChunksCached" without holding lock "MediaTrack.mutex". Elsewhere, "MediaTrack.numberOfFragmentChunksCached" is written to with "MediaTrack.mutex" held 3 out of 3 times (2 of these accesses strongly imply that it is necessary).

_Medium Impact, CWE-366_
_MISSING_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[fragmentcollector_mpd.cpp:9107](undefined//blob/undefined/fragmentcollector_mpd.cpp#L9107)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0385e6553fb826b41843f8a1962dbb38
PRESENT
-->

# Coverity Issue - Unchecked return value
Calling "GetCurrentTextTrack" without checking return value (as is done elsewhere 1 out of 1 times).

_Medium Impact, CWE-252_
_CHECKED_RETURN_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[priv_aamp.cpp:10163](undefined//blob/undefined/priv_aamp.cpp#L10163)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
b7e9402d5325cfffffc304da46f07d5d
PRESENT
-->

# Coverity Issue - Double unlock
"PushNextFragment" unlocks "pMediaStreamContext->mutex" while it is unlocked.

_Medium Impact, CWE-765_
_LOCK_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[fragmentcollector_mpd.cpp:9048](undefined//blob/undefined/fragmentcollector_mpd.cpp#L9048)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
59c199fc71ff333a110953b523743fb8
PRESENT
-->

# Coverity Issue - Unchecked return value
Calling "GetCurrentAudioTrack" without checking return value (as is done elsewhere 2 out of 2 times).

_Medium Impact, CWE-252_
_CHECKED_RETURN_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[priv_aamp.cpp:10043](undefined//blob/undefined/priv_aamp.cpp#L10043)

dnfaulkner commented on priv_aamp.cpp:
```diff
@@ -5523,6 +5512,26 @@ void PrivateInstanceAAMP::TuneHelper(TuneType tuneType, bool seekWhilePaused)
+				double flushPosition = (mMediaFormat == eMEDIAFORMAT_PROGRESSIVE) ? updatedSeekPosition : mpStreamAbstractionAAMP->GetFirstPTS();
+				if (mMediaFormat == eMEDIAFORMAT_DASH)
+				{
+					sink->Flush(flushPosition, rate, false);
```
Maybe it would be useful to add a comment to explain why shouldTearDown is false for DASH?

anjali-syna replied:
Thanks for the comment. Code is modified now.

dnfaulkner reviewed (commented):
One small comment

anjali-syna committed:
VPLAY-10090 [ENT-4082] Audio loss in Trick to play transition in L2 tests

Reason for change: Audio is lost while performing live pause
 and fast forward to live using LLD test asset in L2 test environment.
 This is because in TuneHelper,  Configure is happening after Flush() call
 with proper seek position. This results in pileline loosing the flush and
 seek, thus affecting AV sync.

Changes:
* Perform Flush after Configure in TuneHelper
* Also, call flush with shouldTearDown false, because in case of a new tune,
  pipeline might not be in a playing/paused state which causes Flush() to
  destroy the pipeline. This has to be avoided.
* Adding new unit test:TuneHelperWithAampTsbConfigureFlushSequence

Test Procedure: Run L2 tests

Risks: Low

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

anjali-syna committed:
VPLAY-10090 [ENT-4082] Audio loss in Trick to play transition in L2 tests

Reason for change: Audio is lost while performing fast forward to live
 in L2 test environment. This is because in TuneHelper,  Configure is
 happening after Flush() call with proper seek position.
 This results in pipeline loosing the flush and seek, thus affecting AV sync.

Changes:
* Perform Flush after Configure in TuneHelper
* Also, call flush with shouldTearDown false, because in case of a new tune,
  pipeline might not be in a playing/paused state which causes Flush() to
  destroy the pipeline. This has to be avoided.
* Adding new unit test:TuneHelperWithAampTsbConfigureFlushSequence

Test Procedure: Refer JIRA ticket

Risks: Low

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

jfagunde reviewed (approved):
Good job!

anjali-syna committed:
VPLAY-10090 [ENT-4082] Audio loss in Trick to play transition in L2 tests

Reason for change: Audio is lost while performing fast forward to live
 in L2 test environment. This is because in TuneHelper,  Configure is
 happening after Flush() call with proper seek position.
 This results in pipeline loosing the flush and seek, thus affecting AV sync.

Changes:
* Perform Flush after Configure in TuneHelper
* Also, call flush with shouldTearDown false, because in case of a new tune,
  pipeline might not be in a playing/paused state which causes Flush() to
  destroy the pipeline. This has to be avoided.
* Adding new unit test:TuneHelperWithAampTsbConfigureFlushSequence

Test Procedure: Refer JIRA ticket

Risks: Low

Signed-off-by: anjali-syna <206662904+anjali-syna@users.noreply.github.com>

Vinish100 commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -1564,6 +1564,8 @@ bool InterfacePlayerRDK::Flush(double position, int rate, bool shouldTearDown, b
 		MW_LOG_ERR("Seek failed");
 		SetPendingSeek(true);
+		//Save the updated seek position
+		SetSeekPosition(position);
```
This is not required. SeekPosition is set at the beginning, hence only need to set the pendingSeek flag
You should add a else block for shouldTearDown in line 1529 and mark SetPendingSeek(true);, so that flush would happen before the first data is injected in SendGstEvents.

anjali-syna replied:
This is actually the fix for the L2 failures i mentioned yesterday. Even though seek position is set at the beginning, line 1587 is altering it for certain cases and the pending seek will miss this updated position if we don't call SetSeekPosition at line 1568. Adding line 1568 solves the L2 failures seen in 4006 and 4007 tests.

Vinish100 reviewed (approved):
Looks good!
